File: app/src/main/java/android/print/PdfConverter.java
package android.print;

import android.content.Context;
import android.os.Handler;
import android.os.ParcelFileDescriptor;
import android.util.Log;
import android.webkit.WebView;

import java.io.File;

public class PdfConverter implements Runnable {

    private PdfConverter(){}
    private Context mContext = null;
    private WebView mwebView= null;
    private File mPdfFile = null;
    private PrintAttributes defaultPrintAttrs = null;
    private PrintAttributes pdfPrintAttrs = null;
   // get() = if (field != null) field else defaultPrintAttrs
    private Boolean mIsCurrentlyConverting = false;
    private OnComplete mOnComplete = null;
    ParcelFileDescriptor parcelFileDescriptor=null;

    private static String TAG = "PdfConverter";
    private static PdfConverter sInstance = null;


    public static synchronized PdfConverter getInstance() {
        if (sInstance == null)
            sInstance = new PdfConverter();

        return sInstance;
    }

    public interface OnComplete {
        void onWriteComplete();
        void onWriteFailed();
    }

    private PrintAttributes getPdfPrintAttrs() {
        if(pdfPrintAttrs!=null){
        }else{
            pdfPrintAttrs = getDefaultPrintAttrs();
        }
        return pdfPrintAttrs;
    }

    private ParcelFileDescriptor getParcelFileDescriptor() {
        try {
            mPdfFile.createNewFile();
            return ParcelFileDescriptor.open(mPdfFile, ParcelFileDescriptor.MODE_TRUNCATE | ParcelFileDescriptor.MODE_READ_WRITE);
        } catch ( Exception e) {
            Log.d(TAG, "Failed to open ParcelFileDescriptor", e);
        }

        return null;
    }

    private PrintAttributes getDefaultPrintAttrs() {
        return new PrintAttributes.Builder()
                .setMediaSize(PrintAttributes.MediaSize.ISO_A4)
                .setResolution(new PrintAttributes.Resolution("RESOLUTION_ID", "RESOLUTION_ID", 600, 600))
                .setMinMargins(PrintAttributes.Margins.NO_MARGINS)
                .build();
    }

    @Override
    public void run() {
        final PrintDocumentAdapter documentAdapter =mwebView.createPrintDocumentAdapter("doc");

        documentAdapter.onLayout(null, getPdfPrintAttrs(), null,new PrintDocumentAdapter.LayoutResultCallback() {
            @Override
            public void onLayoutFinished(PrintDocumentInfo info, boolean changed) {
                super.onLayoutFinished(info, changed);
                documentAdapter.onWrite(new PageRange[]{PageRange.ALL_PAGES}, getParcelFileDescriptor(), null, new PrintDocumentAdapter.WriteResultCallback() {
                    @Override
                    public void onWriteFinished(PageRange[] pages) {
                        super.onWriteFinished(pages);
                        mOnComplete.onWriteComplete();
                        destroy();
                    }

                    @Override
                    public void onWriteFailed(CharSequence error) {
                        super.onWriteFailed(error);
                        mOnComplete.onWriteFailed();

                    }

                    @Override
                    public void onWriteCancelled() {
                        super.onWriteCancelled();
                    }
                });
            }
        },null);
    }
    private void runOnUiThread(Runnable runnable) {
        Handler handler =new Handler(mContext.getMainLooper());
        handler.post(runnable);
    }

    public void convert(Context context,WebView htmlString,File file,OnComplete onComplete) throws Exception {
        if (context == null)
            throw new Exception("context can't be null");
        if (htmlString == null)
            throw new Exception("htmlString can't be null");
        if (file == null)
            throw new Exception("file can't be null");

        if (mIsCurrentlyConverting)
            return;

        mContext = context;
        mwebView = htmlString;
        mPdfFile = file;
        mIsCurrentlyConverting = true;
        mOnComplete = onComplete;
        runOnUiThread(this);
    }

    private void destroy() {
        mContext = null;
        mwebView = null;
        mPdfFile = null;
        pdfPrintAttrs = null;
        mIsCurrentlyConverting = false;
        mwebView = null;
        mOnComplete = null;
    }
}

File: app/src/main/java/com/jolly/imagecompression/print/PdfConverter.kt
package android.print

import android.content.Context
import android.os.Handler
import android.os.ParcelFileDescriptor
import android.util.Log
import android.webkit.WebView
import java.io.File

internal class PdfConverter private constructor() : Runnable {

    private var mContext: Context? = null
    private var mwebView: WebView? = null
    private var mPdfFile: File? = null
    private var pdfPrintAttrs: PrintAttributes? = null
        get() = if (field != null) field else defaultPrintAttrs
    private var mIsCurrentlyConverting: Boolean = false
    private var mOnComplete: OnComplete? = null

    private val outputFileDescriptor: ParcelFileDescriptor?
        get() {
            try {
                mPdfFile!!.createNewFile()
                return ParcelFileDescriptor.open(mPdfFile, ParcelFileDescriptor.MODE_TRUNCATE or ParcelFileDescriptor.MODE_READ_WRITE)
            } catch (e: Exception) {
                Log.d(TAG, "Failed to open ParcelFileDescriptor", e)
            }

            return null
        }

    private val defaultPrintAttrs: PrintAttributes?
        get() = PrintAttributes.Builder()
            .setMediaSize(PrintAttributes.MediaSize.ISO_A4)
            .setResolution(PrintAttributes.Resolution("RESOLUTION_ID", "RESOLUTION_ID", 600, 600))
            .setMinMargins(PrintAttributes.Margins.NO_MARGINS)
            .build()

    override fun run() {


                    val documentAdapter =mwebView!!.createPrintDocumentAdapter("doc")

                    documentAdapter.onLayout(null, pdfPrintAttrs, null, object : PrintDocumentAdapter.LayoutResultCallback() {
                        override fun onLayoutFinished(info: PrintDocumentInfo?, changed: Boolean) {
                            super.onLayoutFinished(info, changed)

                            documentAdapter.onWrite(arrayOf(PageRange.ALL_PAGES), outputFileDescriptor, null, object : PrintDocumentAdapter.WriteResultCallback() {
                                override fun onWriteFinished(pages: Array<PageRange>) {
                                    mOnComplete?.onWriteComplete()
                                    destroy()
                                }

                                override fun onWriteFailed(error: CharSequence?) {
                                    super.onWriteFailed(error)
                                    mOnComplete?.onWriteFailed()
                                }
                            })

                        }
                    }, null)



       // mWebView!!.loadDataWithBaseURL("", mwebView, "text/html", "UTF-8", null)
    }

    @Throws(Exception::class)
    fun convert(context: Context?, htmlString: WebView?, file: File?, onComplete: OnComplete?) {
        if (context == null)
            throw Exception("context can't be null")
        if (htmlString == null)
            throw Exception("htmlString can't be null")
        if (file == null)
            throw Exception("file can't be null")

        if (mIsCurrentlyConverting)
            return

        mContext = context
        mwebView = htmlString
        mPdfFile = file
        mIsCurrentlyConverting = true
        mOnComplete = onComplete
        runOnUiThread(this)
    }

    private fun runOnUiThread(runnable: Runnable) {
        val handler = Handler(mContext!!.mainLooper)
        handler.post(runnable)
    }

    private fun destroy() {
        mContext = null
        mwebView = null
        mPdfFile = null
        pdfPrintAttrs = null
        mIsCurrentlyConverting = false
        mwebView = null
        mOnComplete = null
    }

    companion object {

        private const val TAG = "PdfConverter"
        private var sInstance: PdfConverter? = null

        val instance: PdfConverter
            @Synchronized get() {
                if (sInstance == null)
                    sInstance = PdfConverter()

                return sInstance!!
            }

        interface OnComplete {
            fun onWriteComplete()
            fun onWriteFailed()
        }
    }
}